### DISCUSSION TITLE
Is it possible to get notified once all tokio spawned tasks are finished?

### DISCUSSION THREAD

**Discussion by Folyd** [tokio-rs/tokio#5539] [Q&A]
I'm developing https://github.com/zineland/zine, when I run `zine build` it spawns 
a lot of tasks (including `spawn_blocking()`), however, some watch channels will 
fail due to the process exit without waiting for all tasks to finish. 

```
$ zine build
Build cost: 57ms
Build success! The build directory is `build`.
thread 'tokio-runtime-worker' panicked at 'URL preview watch channel receive failed.: RecvError(())', /Users/xxx/.cargo/registry/src/github.com-1ecc6299db9ec823/zine-0.12.1/src/code_blocks/mod.rs:72:30
```

Those tasks spawned in multiple places, I can't simply use `join_all()` to wait 
all it is finished. I hope I can get notified once all spawned tasks are finished. 

Here is a simple snippet to demonstrate my idea:

```rust
runtime.wait_all_task_finished().await;
std::process::exit(0)
```

**Comment by Darksonn** [comment]
Tokio doesn't come with a built-in way to do this, but check out [this page on graceful shutdown](https://tokio.rs/tokio/topics/shutdown).

**Comment by Folyd** [comment]
Thanks, Alice. Good to know the mpsc channel way. 
However, I think it is nearly unfeasible to adopt it at scale. 
It is not uncommon to spawn tasks in multiple places and nested deeply.
It would be better if tokio can capable of such a feature.

What do you think?

**Reply by Darksonn** [reply]
I think there is already a feature request for this.

**Reply by Folyd** [reply]
After a quick search, I found no related feature request issue. Did I miss it?
